### PR TITLE
Allow caller to specify the prefix in a more flexible way when performing backup

### LINKED ISSUE #20335: allow caller to specify the prefix in a more flexible way when performing backup or restore
now that we have

- backup API. see https://github.com/scylladb/scylladb/blob/e4b213f041b131f38a9d782c67152e1203bd3a7e/api/api-doc/storage_service.json#L750
- restore API. see #20305 (still pending on review)

they work fine in the sense that the related tests pass, and the restore API is able to restore the snapshot backup'ed using the backup API. but scylla-manager (SM for short) uses a different naming convention for the paths of objects stored in the remote.

- scylladb uses: `$bucket/$table_name/$snapshot_name/$sstable_component`
- SM uses: `$dc:$provider:$location_path/backup/sst/cluster/$cluster_id/dc/$dc_id/node/$node_id/keyspace/$keyspace/table/$table_name/$snapshot_version/$sstable_component`

in comparison with SM's hierarchical struture, scylladb's arrangement is more flattened.  but SM does this for a couple reasons:

- it wants to avoid name conflicts: as the same node could upload the sstable with the same sstable with the same name, but from different snapshot
- so SM can list with a given prefix to resume a backup operation
- so SM can list with a given prefix for dedup -- it keeps track of the sstable files using a manifest file with the same prefix.

also, instead of calling the "restore" API,  SM will have to download the backup'ed sstables using agent so it can perform load-and-stream for restoring vnode-based keyspaces.

so the backup and restore APIs need to allow the caller to specify both the bucket *and* the prefix, or at least the pattern of the prefix. this would allow SM to keep using the existing techniques to keep track of the progress and to restore a vnode keyspace.

this issue is created to track the design of the updated APIs allowing the caller to pass the prefix / or the pattern of prefix when performing backup and restore.

regarding the backup API, i am proposing following interface:

```json
{
          "path":"/storage_service/backup",
          "operations":[
              {
                  "method":"POST",
                  "summary":"Starts copying SSTables from a specified keyspace to a designated bucket in object storage",
                  "type":"string",
                  "nickname":"start_backup",
                  "produces":[
                      "application/json"
                  ],
                  "parameters":[
                      {
                          "name":"endpoint",
                          "description":"ID of the configured object storage endpoint to copy sstables to",
                          "required":true,
                          "allowMultiple":false,
                          "type":"string",
                          "paramType":"query"
                      },
                      {
                          "name":"prefix",
                          "description":"The prefix of the objects for the backuped sstables",
                          "required":true,
                          "allowMultiple":false,
                          "type":"string",
                          "paramType":"query"
                      },
                      {
                          "name":"keyspace",
                          "description":"Name of a keyspace to copy sstables from",
                          "required":true,
                          "allowMultiple":false,
                          "type":"string",
                          "paramType":"query"
                      },
                      {
                          "name":"snapshot",
                          "description":"Name of a snapshot to copy sstables from",
                          "required":false,
                          "allowMultiple":false,
                          "type":"string",
                          "paramType":"query"
                      }
                  ]
              }
          ]
      },
```
where, "bucket" is replaced with "prefix". the same applies to the "restore" API:

```json
{
          "path":"/storage_service/restore",
          "operations":[
              {
                  "method":"POST",
                  "summary":"Starts copying SSTables from a designated bucket in object storage to a specified keyspace",
                  "type":"string",
                  "nickname":"start_restore",
                  "produces":[
                      "application/json"
                  ],
                  "parameters":[
                      {
                          "name":"endpoint",
                          "description":"ID of the configured object storage endpoint to copy SSTables from",
                          "required":true,
                          "allowMultiple":false,
                          "type":"string",
                          "paramType":"query"
                      },
                      {
                          "name":"prefix",
                          "description":"Prefix of objects from which SSTables are restored",
                          "required":true,
                          "allowMultiple":false,
                          "type":"string",
                          "paramType":"query"
                      },
                      {
                          "name":"snapshot",
                          "description":"Name of a snapshot to copy SSTables from",
                          "required":true,
                          "allowMultiple":false,
                          "type":"string",
                          "paramType":"query"
                      },
                      {
                          "name":"keyspace",
                          "description":"Name of a keyspace to copy SSTables to",
                          "required":true,
                          "allowMultiple":false,
                          "type":"string",
                          "paramType":"query"
                      },
                      {
                          "name":"table",
                          "description":"Name of a table to copy SSTables to",
                          "required":false,
                          "allowMultiple":false,
                          "type":"string",
                          "paramType":"query"
                      }
                  ]
              }
          ]
      }
```

### PR DESCRIPTION
**Pull request by kreuzerkrieg**
Fixes #20335 by allowing the caller to pass the prefix when performing backup and restore

### CODE REVIEW

**Comment by mykaul** [comment]
We use 'Fixes:' or 'Refs:' , not 'Addresses' - see https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests#linking-a-pull-request-to-an-issue

**Comment by kreuzerkrieg** [comment]
@Michal-Leszczynski I guess you are interested to be aware of this PR

File: test/object_store/test_backup.py
Comment (tchaikov) [inline test/object_store/test_backup.py:108]: might want to replace `'foo/bar/baz'` with `prefix`.

File: test/object_store/test_backup.py
Comment (tchaikov) [inline test/object_store/test_backup.py:122]: ditto.

File: test/object_store/test_backup.py
Comment (tchaikov) [inline test/object_store/test_backup.py:167]: i'd suggest use a variable for the magic number here.

**Review by tchaikov — CHANGES_REQUESTED** [review]
the commit title and its body reads:
```
Make backup/restore compatible with Scylla Manager

# Conflicts:
#	tools/scylla-nodetool.cc
```

i think we should be more specific in the title,  put the rationales in the commit message, and remove the "Conflicts" part.  something like:

```
treewide: add "prefix" parameter to backup and restore API

/// put more context here
```

**Review by regevran — APPROVED** [review]

**Review by regevran — CHANGES_REQUESTED** [review]
This PR brings API changes and therefore must be correlated with Scylla Manager.

File: db/snapshot-ctl.cc
Comment (xemul) [inline db/snapshot-ctl.cc:141]: Please, keep it multiline

**Review by xemul — APPROVED** [review]
LGTM, modulo minor comment about formatting

File: test/nodetool/test_restore.py
Comment (kreuzerkrieg) [reply test/nodetool/test_restore.py:22]: typo, will fix

File: db/snapshot-ctl.cc
Comment (kreuzerkrieg) [reply db/snapshot-ctl.cc:141]: will do

File: docs/dev/object_storage.md
Comment (kreuzerkrieg) [reply docs/dev/object_storage.md:111]: Done

**Review by tchaikov — APPROVED** [review]
lgtm

File: docs/operating-scylla/nodetool-commands/restore.rst
Comment (kreuzerkrieg) [reply docs/operating-scylla/nodetool-commands/restore.rst:28]: Done

**Comment by avikivity** [comment]
> resolves #20335 by allowing the caller to pass the prefix when performing backup and restore

Please use 'Fixes #nnnnn' on a line by itself so it's unambiguous for our automation.

**Comment by tchaikov** [comment]
@kreuzerkrieg hi Ernest, apart from Avi's comment on the cover letter of the PR. i'd like to remind explain more on our conventions on sending pull requests. we don't fix previous commit with another commit in the same PR. for instance, it is not encouraged to include following sequence in a single PR:

1. make a change A, 
2. update A to better one by rewriting all of it, named A'

or 

1. make a change A and B
2. revert B

but IIUC, it's totally fine to have 

1. make a change A which lays down the general structure, or do the ground work for succeeding changes.
2. add B which uses the the facilities / framework provided by A
3. add C which uses the the facilities / framework provided by A and/or B

when it comes to this very PR, we are at a spot where the proposed API (#20335) was superseded by a new one (#20461) even before the PR implementing the proposal has not been merged. unless i misunderstood, we had the consensus on merging it over the meeting yesterday.  because the "restore" API will not be compatible with the "backup" API if we only change the latter, we cannot just have one of them in a PR, otherwise the test fails. there are three options:

1. keep and merge this PR as we agreed over the meeting by implementing #20335 (cc @xemul )
2. keep this PR, but update the "restore" API, so and pass a list of names of sstables. so it partially implements #20335 and partially implements #20461
3. abandon this PR or repurpose it for #20461

i'd suggest go with option 1 even a part of the change won't exist in the final version. otherwise,  we could consider option 3 after we have nail down the details of the APIs, like how do we pass the manifest and list, and the parameter names.

File: test/object_store/test_backup.py
Comment (tchaikov) [inline test/object_store/test_backup.py:156]:
```suggestion
    prefix = snap_name
```

**Review by tchaikov — APPROVED** [review]
lgtm modulo a nit.

File: sstables_loader.hh
Comment (xemul) [inline sstables_loader.hh:90]: I see the change in method signature, but don't see the updated callers. Does it compile at all?

Comment (kreuzerkrieg) [reply sstables_loader.hh]: Reverted

**Comment by Michal-Leszczynski** [comment]
@kreuzerkrieg how is prefix going to work when the current backup API has keyspace granularity?
Would it mean that all sstables from given keyspace are going to be stored in the same dir in backup location?
This wouldn't be compatible withe the current backup dir layout and would introduce problems when handling sstables with integer ID generations.

Are we planning to support the table param for backup API as well (like in restore)?

**Review by regevran — COMMENTED** [review]
Let's hold this PR until we have a clear documentation of Scylla's backup and restore API.
We need to produce this document with @Michal-Leszczynski.

**Comment by kreuzerkrieg** [comment]
> @kreuzerkrieg how is prefix going to work when the current backup API has keyspace granularity? Would it mean that all sstables from given keyspace are going to be stored in the same dir in backup location? This wouldn't be compatible withe the current backup dir layout and would introduce problems when handling sstables with integer ID generations.
> 
> Are we planning to support the table param for backup API as well (like in restore)?

Not sure I'm following, what prevents us from including the keyspace in the prefix? Like it is mentioned [here](https://github.com/scylladb/scylladb/issues/20335#issuecomment-2323247331)

**Comment by tchaikov** [comment]
@Michal-Leszczynski i would prefer keeping the discussion in a single place. if you don't mind, i am answering your questions at #20335.

**Comment by tchaikov** [comment]
@xemul @regevran i think we already have reached a consensus on having a green light for this PR, right? so the answer is yes, may we merge this? i'd like to add the "table" support as a follow-up.

**Comment by tchaikov** [comment]
@xemul Hi Pavel, what do you think ? ^

**Comment by kreuzerkrieg** [comment]
@scylladb/scylla-maint hello maintainers, could you help queue this change?

**Comment by denesb** [comment]
@kreuzerkrieg please drop the merge commit, rebase your PR on top of latest master instead.

**Comment by denesb** [comment]
Also, how will this fix work if only backup API has prefix param? Don't we need it for restore too?

**Comment by kreuzerkrieg** [comment]
> @kreuzerkrieg please drop the merge commit, rebase your PR on top of latest master instead.

Super weird, I dont have such a commit in my git log, will figure out whats going on

**Comment by kreuzerkrieg** [comment]
> Also, how will this fix work if only backup API has prefix param? Don't we need it for restore too?

The prefix provided in backup should be matching restore expectations. In any case, this solution is only the first step toward more general solution

**Review by xemul — APPROVED** [review]

**Comment by tchaikov** [comment]
@scylladb/scylla-maint hello maintainers, could you help queue this change?

**Pull request closed without merging**
